### PR TITLE
WIP Added onboarding

### LINKED ISSUE #64: Add onboarding on the first run
http://atlaskit.atlassian.com/packages/core/onboarding

- tell the user what to enter in the room field
- expand and explain settings

### PR DESCRIPTION
**Pull request by akshitkrnagpal**
Resolves #64 

TODO
- [x] Fix onboarding. 🤕 
- [x] Welcome Modal with actions (Start Tour and Skip)
- [x] Add onboarding modal image
- [x] Persist

### CODE REVIEW

File: app/features/onboarding/constants.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+export const onboardingSteps = [
+    'conference-url',
+    'settings-drawer-button',
+    'name-setting',
+    'email-setting',
+    'server-setting'
+];
+
+export const onboardingComponents = {
```
Comment (akshitkrnagpal) [inline app/features/onboarding/constants.js]: As we have to persist which onboarding components have been shown already, we can't directly use components. Hence we have to map them with strings.

File: app/features/app/components/App.js [hunk 2]
```diff
@@ -50,3 +51,5 @@ export default class App extends Component<*> {
         );
     }
 }
+
+export default connect()(App);
```
Comment (saghul) [inline app/features/app/components/App.js]: Is this needed now?

File: app/features/onboarding/actionTypes.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+/**
+ * The type of (redux) action that hides active onboarding.
+ *
+ * {
+ *     type: HIDE_ACTIVE_ONBOARDING,
+ *     parentComponent: string
```
Comment (saghul) [inline app/features/onboarding/actionTypes.js]: How about calling this "section" instead?

File: app/features/welcome/components/Welcome.js [hunk 4]
```diff
@@ -104,12 +128,31 @@ class Welcome extends Component<Props, State> {
...
             </Page>
         );
     }
 
+    /**
+     * Render active onboarding component.
+     *
+     * @returns {void}
+     */
+    _renderActiveOnboardingComponent() {
```
Comment (saghul) [inline app/features/welcome/components/Welcome.js]: I think this could be put into a `OnboardingContainer` or just `Onboarding` component, which gets a prop with the section name. That way it can be reused in both locations.

File: app/features/settings/components/SettingsDrawer.js [hunk 5]
```diff
@@ -63,6 +74,24 @@ class SettingsDrawer extends Component<Props, *> {
...
+     *
+     * NOTE: It autonatically checks if the onboarding is shown or not.
+     *
+     * @returns {void}
+     */
+    componentDidUpdate() {
+        if (this.props.isOpen) {
+
+            // TODO - Find a better way for this.
+            // Delay for 1 second to let drawer open.
```
Comment (saghul) [inline app/features/settings/components/SettingsDrawer.js]: Is one full second necessary? Or just a 100ms setTimeout would do?

File: app/features/settings/components/SettingsDrawer.js [hunk 6]
```diff
@@ -63,6 +74,24 @@ class SettingsDrawer extends Component<Props, *> {
...
 
+    /**
+     * Start Onboarding once component is mounted.
+     *
+     * NOTE: It autonatically checks if the onboarding is shown or not.
+     *
+     * @returns {void}
+     */
+    componentDidUpdate() {
+        if (this.props.isOpen) {
```
Comment (saghul) [inline app/features/settings/components/SettingsDrawer.js]: can you check if `!prevProps.isOpen` here? (just playing safe)

File: app/features/settings/components/SettingsDrawer.js (hunk 5 above)
Comment (akshitkrnagpal) [reply app/features/settings/components/SettingsDrawer.js]: Drawer animation is 220ms. So I guess we at least need that much.
On fast devices, 250ms works just fine. But on a little slow device 300ms sometimes does it correctly and sometimes the first onboarding created the problem which I mentioned earlier.

WDYT, how much should we set?

File: app/features/onboarding/actionTypes.js (hunk 3 above)
Comment (akshitkrnagpal) [reply app/features/onboarding/actionTypes.js]: Done

File: app/features/welcome/components/Welcome.js (hunk 4 above)
Comment (akshitkrnagpal) [reply app/features/welcome/components/Welcome.js]: Nice catch.
I tried that when we didn't have sections for onboarding but I guess I just forgot. :smile:

File: app/features/settings/components/SettingsDrawer.js (hunk 6 above)
Comment (akshitkrnagpal) [reply app/features/settings/components/SettingsDrawer.js]: Done

File: app/features/settings/components/SettingsDrawer.js (hunk 5 above)
Comment (saghul) [reply app/features/settings/components/SettingsDrawer.js]: Let's do 500, and please put that 220 in the comment so we remember :-)

File: app/features/app/components/App.js [hunk 7]
```diff
@@ -50,3 +50,5 @@ export default class App extends Component<*> {
         );
     }
 }
+
+export default App;
```
Comment (saghul) [inline app/features/app/components/App.js]: Can you please undo this change altogether? It's just not needed so it only adds "nose" to the PR.

**Comment by saghul** [comment]
Looking good @akshitkrnagpal ! Just those 2 items missing and we got this! (I'll make copy edits after you're done to avoid force-pushing)

**Comment by akshitkrnagpal** [comment]
@saghul 
I think this is done except Modal Image.

**Comment by akshitkrnagpal** [comment]
- [x] Updated Dependencies
